src/thor/map_matcher.cc

```diff
@@ -133,59 +134,81 @@ MapMatcher::FormPath(meili::MapMatcher* matcher,
+        nodeinfo = tile->node(directededge->endnode());
+        DateTime::seconds_since_epoch(options.date_time(),
+                                      DateTime::get_tz_db().from_index(nodeinfo->timezone()));
+        nodeinfo = nullptr;
+      }
+    }
+  }
 
   // Interpolate match results if using timestamps for elapsed time
   std::list<std::vector<interpolation_t>> interpolations;
   size_t interpolated_index = 0;
   size_t last_interp_index = 0;
   if (use_timestamps) {
     interpolations = interpolate_matches(results, edge_segments, matcher);
-    if (interpolations.size() > 1) {
-      // This means there is a discontinuity. If so, fallback to using costing
+    // This means there is a discontinuity. If so, fallback to using costing
+    if (interpolations.size() != 1)
       use_timestamps = false;
-    }
-    last_interp_index = interpolations.front().back().original_index;
+    else
+      last_interp_index = interpolations.front().back().original_index;
   }
 
+  // Build the path
   for (const auto& edge_segment : edge_segments) {
-
     // Skip edges that are the same as the prior edge
     if (edge_segment.edgeid == prior_edge) {
       continue;
     }
 
     // Get the directed edge
     GraphId edge_id = edge_segment.edgeid;
-    const GraphTile* tile = matcher->graphreader().GetGraphTile(edge_id);
+    matcher->graphreader().GetGraphTile(edge_id, tile);
     directededge = tile->directededge(edge_id);
-
     // Check if connected to prior edge
     if (prior_edge.Is_Valid() && !matcher->graphreader().AreEdgesConnected(prior_edge, edge_id)) {
       disconnected_edges.emplace_back(prior_edge, edge_id);
     }
 
-    // TODO: slight difference in time between route and trace_route
-    if (nodeinfo) {
-      // Get time along the edge, handling partial distance along the first and last edge.
-      // Add the transition cost at the begin node.
-      elapsed +=
-          costing->EdgeCost(directededge, tile, 0) * (edge_segment.target - edge_segment.source) +
-          costing->TransitionCost(directededge, nodeinfo, pred);
-    } else {
-      // Get time along the edge, handling partial distance along the first and last edge
-      elapsed +=
-          costing->EdgeCost(directededge, tile, 0) * (edge_segment.target - edge_segment.source);
+    // get the cost of traversing the node
+    elapsed += !nodeinfo ? Cost{} : costing->TransitionCost(directededge, nodeinfo, pred);
+
+    // Set seconds from beginning of the week
+    uint32_t second_of_week = kInvalidSecondsOfWeek;
+    // have to always compute the offset in case the timezone changes along the path
+    // we could cache the timezone and just add seconds when the timezone doesnt change
+    // we wont have a valid node on the first pass but then again we also wont have any
+    // elapsed time so this would be irrelevant anyway
+    if (origin_epoch != 0 && nodeinfo) {
+      second_of_week = DateTime::second_of_week(origin_epoch, nodeinfo, elapsed.secs);
     }
```

Comment:
this is the meat of the change. on every new path edge we figure out what second of the week it was (if it was a time dependent route) and use that to do the edge costing below. thsi allows the costing to access the rigth speed for the time of day that route was supposed to have been happening